README
NAME

      Sjis - Source code filter for ShiftJIS script

SYNOPSIS

  Usage: A

    In ShiftJIS_script.pl:
      use Sjis;
      use Sjis version;
      use Sjis qw(ord reverse);
      use Sjis version qw(ord reverse);

    And on command prompt:
      perl     ShiftJIS_script.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl58   ShiftJIS_script.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl510  ShiftJIS_script.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl512  ShiftJIS_script.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl64   ShiftJIS_script.pl  wild*  *card  and  '*quote*'  on MSWin32

  Usage: B

    On command prompt:
      C:\>perl  Sjis.pm  ShiftJIS_script.pl  >  Escaped_script.pl.e

    And
      C:\>perl  Escaped_script.pl.e

DESCRIPTION

      This software is "JPerl on the Modern Perl" written by Perl.

      This "JPerl" means "Japanized Perl" or "Japanese Perl".
      So is unrelated to ...
        Mr. Patrick E. Martin's JPerl -- Perl 5 implemented in Java
        Mr. Ben Evans's jPerl -- Perl on the JVM
        Mr. James C. Estill's jperl --- Jamie's PERL scripts for bioinformatics

      This software treats ShiftJIS directly. Therefore, there is not UTF8 flag.

INSTALLATION BY MAKE (for UNIX)

To install this software by make, type the following:

   perl Makefile.PL
   make
   make test
   make install

INSTALLATION WITHOUT MAKE (for DOS like system)

To install this software without make, type the following:

   perl pMakefile.PL    --- pMakefile.PL makes "pmake.bat" only, and ...
   pmake.bat
   pmake.bat test

   pmake.bat install           --- install to current using Perl
   perl58   pmake.bat install  --- install to Perl5.8
   perl510  pmake.bat install  --- install to Perl5.10
   perl512  pmake.bat install  --- install to Perl5.12
   perl64   pmake.bat install  --- install to Perl64

   pmake.bat dist       --- make distribution package
   pmake.bat ptar.bat   --- make perl script "ptar.bat"

PROGRAM COMPOSITION

   Sjis.pm               --- source code filter to escape ShiftJIS
   Esjis.pm              --- run-time routines for Sjis.pm
   perl58.bat            --- find and run perl5.8  without %PATH% settings
   perl510.bat           --- find and run perl5.10 without %PATH% settings
   perl512.bat           --- find and run perl5.12 without %PATH% settings
   perl64.bat            --- find and run perl64   without %PATH% settings
   warnings.pm_          --- poor warnings.pm
   warnings/register.pm_ --- poor warnings/register.pm

DEPENDENCIES

This software requires perl5.00503 or later.

LICENSE AND COPYRIGHT

This software is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See perlartistic.

This software is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

MAINTAINER

This project was originated by INABA Hitoshi <ina@cpan.org>.

ShiftJIS Encoding

ShiftJIS encoding is the most common encoding method used on Japanese
computers, developed by Microsoft Corporation.

ShiftJIS is sometimes referred to as MS Kanji (Microsoft Kanji) or SJIS
(an abbreviation for ShiftJIS).

Historically, ShiftJIS was so named because of the way the code positions
for two-octet characters shifted around the code positions for half-width
katakana -- Japanese computer users originally used only half-width katakana,
so ShiftJIS was developed in order to maintain backword compatibility.

A two-octet-per-character mode in ShiftJIS is initiated with a octet having
a hexadecimal value ranging between 0x81-0x9F or 0xE0-0xEF. This octet is
subsequently treated as the first octet of an expected two-octet sequence.
The following (second) octet must have a hexadecimal value ranging between
0x40-0x7E or 0x80-0xFC. Note that the first octet's range falls entirely in
the extended ASCII character set -- in the eight-bit character set with the
high-order bit on. Note that ShiftJIS also encodes half-width katakana and
ASCII/JIS-Roman.

Some definitions (in particular, corporate definitions) of ShiftJIS also
contain encoding blocks for user-defined characters or even an encoded position
for a half-width katakana space character. Such encoding blocks and encoded
positions are not useful if true information interchange is desired, because
they are encoded in such a way that they do not convert to encoded positions
in other Japanese encoding methods (that is, JIS and EUC).

ShiftJIS IN WIKIPEDIA

Shift JIS (2008.02.15 15:02:00 JST). In Wikipedia: The Free Encyclopedia.
Retrieved from
http://en.wikipedia.org/wiki/Shift_JIS

Shift JIS (also SJIS, MIME name Shift_JIS) is a character encoding
for the Japanese language originally developed by a Japanese company
called ASCII Corporation in conjunction with Microsoft and standardized
as JIS X 0208 Appendix 1. It is based on character sets defined within
JIS standards JIS X 0201:1997 (for the single-byte characters) and

Changes
Revision history for this software.

0.69  2010/12/21 00:00:00
	- support MacOS.
	; beside "escape" by Delight Prescott of MacPerl Power and Ease

0.68  2010/11/09 00:00:00
	- re-correct 244_3open.t and 245_2binmode.t.
	; created by INABA Hitoshi

0.67  2010/11/08 00:00:00
	- correct 244_3open.t and 245_2binmode.t.
	; created by INABA Hitoshi

0.66  2010/11/07 00:00:00
	- emulate two-argument binmode() on perl5.005.
	- emulate three-argument open() on perl5.005.
	- dummy functions of utf8.pm and bytes.pm.
	- correct write locking routine.
	- bundle warnings.pm_ and warnings/register.pm_.
	; created by INABA Hitoshi

0.65  2010/09/16 00:00:00
	- check code page on DOS-like system before test.
	; created by INABA Hitoshi

0.64  2010/09/14 00:00:00
	- support SjisTk module.
	- support \l, \u, lcfirst and ucfirst.
	- add test of Latin1.
	; created by INABA Hitoshi

0.63  2010/07/04 00:00:00
	- add test of Big5Plus, Big5HKSCS, GB18030 and UHC.
	; created by INABA Hitoshi

0.62  2010/07/03 00:00:00
	- redefine \b and \B.
	; created by INABA Hitoshi

0.61  2010/06/27 00:00:00
	- fix OldUTF8.
	- initialize tr variables on demand.
	- fix Esjis::lc and Esjis::uc's bug.
	; created by INABA Hitoshi

0.60  2010/06/16 00:00:00
	- back *.pm files from /lib.
	; created by INABA Hitoshi

0.59  2010/06/14 00:00:00
	- move *.pm files to /lib.
	- fix Makefile.PL.
	- fix m/[$@]/ to m/[\$\@]/.
	- add HP15 and INFORMIXV6ALS test.
	- support Big5-HKSCS.
	; created by INABA Hitoshi

0.58  2010/06/07 00:00:00
	- fix escapeshellcmd.
	; created by INABA Hitoshi

0.57  2010/06/02 00:00:00
	- do BEGIN { unshift @INC, '/Perl/site/lib/Sjis'; } in Esjis.pm.
	- support 'no Module'.
	- remove special routine of Tk module.
	- test does only one version of perl.
	- support TAP(Test Anything Protocol).
	- add ShiftJIS::Regexp's test.
	- rename Makefile.PL to pMakefile.PL(Poor Makefile.PL) for DOS like system.
	- standard Makefile.PL for UNIX.
	- warn if chdir "\x5c" ended path on $^W is true.
	; created by INABA Hitoshi

0.56  2010/04/21 00:00:00
	- add JPerl's test.
	- fix [^]] in regexp from <'^' then ']'> to <not ']'>.
	- [] in regexp make die "unmatched [] in regexp at ...".
	- support \b and \B in regexp.
	; created by INABA Hitoshi

0.55  2010/04/16 00:00:00
	- test.pl support perl5.012.
	- support when as a statement modifier.
	- support := operator.
	; created by INABA Hitoshi

0.54  2010/04/02 00:00:00
	- fix /i modifier with character class '[ccc]' in regexp.
	- support old UTF-8.
	; created by INABA Hitoshi


make-gbk.pl
######################################################################
#
# make-gbk.pl - Source code filter for making source code filter to escape GBK
#
# Copyright (c) 2008, 2009 INABA Hitoshi <ina@cpan.org>
#
######################################################################

use strict;
use Fcntl;
use Symbol;

for my $io (
    ['Esjis.pm' => 'Egbk.pm'],
    ['Sjis.pm'  => 'GBK.pm'],
) {
    my $i = $io->[0];
    my $o = $io->[1];

    print STDERR "$i --> $o\n";

    my $fh_i = Symbol::gensym();
    sysopen($fh_i, $i, O_RDONLY)                     || die "Can't open file: $i\n";

    my $fh_o = Symbol::gensym();
    sysopen($fh_o, $o, O_WRONLY | O_TRUNC | O_CREAT) || die "Can't open file: $o\n";
    binmode $fh_o;

    while (<$fh_i>) {
        s/\\x81-\\x9F\\xE0-\\xFC/\\x81-\\xFE/g;
        s/\(\?:8\[1-9A-Fa-f\]\|9\[0-9A-Fa-f\]\|\[Ee\]\[0-9A-Fa-f\]\|\[Ff\]\[0-9A-Ca-c\]\)/(?:8[1-9A-Fa-f]|[9A-Ea-e][0-9A-Fa-f]|[Ff][0-9A-Ea-e])/g;
        s/\(\?:20\[1-7\]\|2\[123\]\[0-7\]\|3\[4-6\]\[0-7\]\|37\[0-4\]\)/(?:2[0-7][1-7]|3[0-7][0-6])/g;
        s/\\x40-\\x7E\\x80-\\xFC/\\x40-\\x7E\\x80-\\xFE/g;
        s/\bEsjis\b/Egbk/g;
        s/ShiftJIS/GBK/g;
        s/Sjis/GBK/g;

        print {$fh_o} $_;
    }

    close $fh_i;
    close $fh_o;
}

__END__

=pod

=head1 NAME

make-gbk.pl - Source code filter for making source code filter to escape GBK

=head1 SYNOPSIS

perl make-gbk.pl

=head1 DESCRIPTION

This utility makes GBK.pm from Sjis.pm for escape GBK script.

=head1 BUGS AND LIMITATIONS

Please patches and report problems to author are welcome.

=head1 AUTHOR

INABA Hitoshi E<lt>ina@cpan.orgE<gt>

This project was originated by INABA Hitoshi.
For any questions, use E<lt>ina@cpan.orgE<gt> so we can share
this file.

=head1 LICENSE AND COPYRIGHT

This software is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

This software is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=head1 GBK IN WIKIPEDIA

GBK (2008.08.24 17:04:00 JST). In Wikipedia: The Free Encyclopedia.
Retrieved from
L<http://en.wikipedia.org/wiki/GBK>

GBK is an extension of the GB2312 character set for simplified Chinese
characters, used in the People's Republic of China.
GB stands for National Standard, while K stands for Extension. GBK not only
extended the old standard GB2312 with Traditional Chinese characters, but
also with Chinese characters that were simplified after the establishment
of GB2312 in 1981. With the arrival of GBK, certain names with characters
formerly unrepresentable, like the "rong" character in former Chinese Premier
Zhu Rongji's name, are now representable.

History
In 1993, the Unicode 1.1 standard was released, including 20,902 characters
used in mainland China, Taiwan, Japan and Korea. Following this, China
released GB13000.1-93, a national standard (guo biao) equivalent of Unicode
1.1.

The GBK character set was defined in 1993 as an extension of GB2312-80,
while also including the characters of GB13000.1-93 through the unused
codepoints available in GB2312. Hence GBK is upward compatible with GB2312.

Microsoft implemented GBK in Windows 95 as Code Page 936. While GBK was
never an official standard, widespread usage of Windows 95 led to GBK
becoming the de facto standard. While GBK included all the Chinese
characters defined in Unicode 1.1 and GB13000.1-93, these standards used
different code tables. The primary reason for its existence was simply to
bridge the gap between GB2312-80 and GB13000.1-93.

In 1995, China National Information Technology Standardization Technical
Committee set down the Chinese Internal Code Specification (simplified
Chinese: GBK Version 1.0, known as GBK 1.0, which is a slight extension of
Codepage 936. The newly added 95 characters were not found in
GB 13000.1-1993, and were provisionally assigned Unicode PUA code points.

Microsoft later added the euro sign to Codepage 936 and assigned the code
0x80 to it. This is not a valid code point in GBK 1.0.

In 2000, the GB18030-2000 standard was released, superseding yet maintaining
compatibility with GBK 1.0. It increased the number of definitions of
Chinese characters and extended the number of possible characters through
the implementation of four-byte character spaces. The subset of GB 18030
consisting of one-byte and two-byte characters is sometimes also referred
to as GBK. Mapping to Unicode has been slightly changed, though, as some
characters are now defined in Unicode. In the most up-to-date form of the
standard, GB 18030-2005, only 14 characters are still mapped to Unicode PUA.

Encoding
A character is encoded as 1 or 2 bytes. A byte in the range 00-7F is a
single byte that means the same thing as it does in ASCII. Strictly speaking,
there are 96 characters and 32 control codes in this range.

A byte with the high bit set indicates that it is the first of 2 bytes.
Loosely speaking, the first byte is in the range 81-FE (that is, never 80
or FF, and the second byte is 40-FE for some areas and 80-FE for others.

More specifically, the following ranges of bytes are defined:

Relationship to other encodings
The areas indicated in the previous section as GBK/1 and GBK/2, taken by
themselves, is simply GB2312-80 in its usual encoding. GB2312, or more
properly the EUC-CN encoding thereof, takes a pair of bytes from the range
A1-FE, like any 94^2 ISO-2022 character set loaded into GR. This corresponds
to the lower-right quarter of the illustration above. However, GB2312 does
not assign any code points to the rows located at AB-B0 and F8-FE, even
though it had staked out the territory.

GBK added extensions to this. You can see that the two gaps were filled in
with user-defined areas.

More significantly, it extended the range of the bytes. Having two-byte
characters in the ISO-2022 GR range gives a limit of 94^2=8,836 possibilities.
Abandoning the ISO-2022 model of strict regions for graphics and control
characters, but retaining the feature of low bytes being 1-byte characters
and pairs of high bytes denoting a character, you could potentially have 128^2
=16,384 positions. GBK takes part of that, extending the range from A1-FE
(94 choices for each byte) to 81-FE (126 choices) for the first byte and 40-FE
(191 choices) for the second byte, for a total of 24,066 positions.

Microsoft's Code Page 936 is generally thought of as being GBK. It has bytes
in the same range, with assignments that seem to match if you compare them.
However, the total number of two-byte code points defined is 21,791 so there
must be some differences --- at the very least, 95 are missing.

GBK's successor, GB18030-2000, uses the remaining range available to the
second byte to further expand the number of possibilities while retaining GBK
as a subset.

=head1 "GBK" IN THIS SOFTWARE

 The "GBK" in this software means widely codeset than general GBK.
This software use two algorithms to handle GBK.

=over 2

=item * ALGORITHM #1

 When the character is taken out of the octet string, it is necessary to
distinguish a single octet character and the double octet character.
The distinction is done only by first octet.

    Single octet code is:
      0x00-0x80 and 0xFF

    Double octet code is:
      First octet   0x81-0xFE
      Second octet  0x00-0xFF (All octet)

    MALFORMED single octet code is:
      0x81-0xFE
      * Final octet of string like first octet of double octet code

See also code table:

         Single octet code

   0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x00-0x80
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8|*| | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F| | | | | | | | | | | | | | | |*| 0xFF
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


                                 Double octet code
            First octet                                     Second octet

   0 1 2 3 4 5 6 7 8 9 A B C D E F                 0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0| | | | | | | | | | | | | | | | |              0|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x00-0xFF
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1| | | | | | | | | | | | | | | | |              1|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2| | | | | | | | | | | | | | | | |              2|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3| | | | | | | | | | | | | | | | |              3|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4| | | | | | | | | | | | | | | | |              4|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5| | | | | | | | | | | | | | | | |              5|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6| | | | | | | | | | | | | | | | |              6|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7| | | | | | | | | | | | | | | | |              7|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8| |*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x81-0xFE    8|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              9|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              A|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              B|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              C|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              D|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              E|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| |              F|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


    *MALFORMED* Single octet code

Final octet of string like first octet of double octet code

Even if malformed, it is not ignored and not deleted automatically.
For example, Esjis::chop function returns this octet.

   0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8| |M|M|M|M|M|M|M|M|M|M|M|M|M|M|M| 0x81-0xFE
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M| |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

=item * ALGORITHM #2

Against algorithm.1, when the range of the character by tr/// is specified, only the
following character codes are effective.

    Single octet code is:
      0x00-0x80 and 0xFF

    Double octet code is:
      First octet   0x81-FE
      Second octet  0x40-0x7E and 0x80-0xFE

For instance, [\x81\x00-\x82\xFF] in script means [\x81\x82][\x40-\x7E\x80-\xFE].

See also code table:

         Single octet code

   0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x00-0x80
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8|*| | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F| | | | | | | | | | | | | | | |*| 0xFF
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


                                 Double octet code
            First octet                                     Second octet

   0 1 2 3 4 5 6 7 8 9 A B C D E F                 0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0| | | | | | | | | | | | | | | | |              0| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1| | | | | | | | | | | | | | | | |              1| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2| | | | | | | | | | | | | | | | |              2| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3| | | | | | | | | | | | | | | | |              3| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4| | | | | | | | | | | | | | | | |              4|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x40-0x7E
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5| | | | | | | | | | | | | | | | |              5|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6| | | | | | | | | | | | | | | | |              6|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7| | | | | | | | | | | | | | | | |              7|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8| |*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x81-0xFE    8|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x80-0xFE
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              9|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              A|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              B|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              C|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              D|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              E|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| |              F|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


=back

=head1 SEE ALSO

perl, Sjis.pm

=cut
